### PR TITLE
[MPS] Migrate `norm` to metal kernel

### PR DESCRIPTION
**Pull request by kurtamohler**
This also adds complex input support.

Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #177328

Performance was measured with the following script:

<details><summary>Click to expand</summary>

```python
import torch
import torch.utils.benchmark as benchmark
import itertools
import statistics

device = "mps"
sizes = [
    (10,),
    (10, 10),
    (1024,),
    (1024, 10),
    (10, 10, 10),
]
dtypes = [torch.float32, torch.float16, torch.bfloat16]
p_values = [1.5, 0]
dim_options = [None, 0, -1]
num_runs = 1000

def run_benchmark(x, p, dim):
    stmt = "torch.norm(x, p=p, dim=dim)"
    timer = benchmark.Timer(
        stmt=stmt,
        globals={"torch": torch, "x": x, "p": p, "dim": dim},
    )
    for _ in range(2):
        result = timer.timeit(num_runs)
    return result.median

results = []

for size, dtype, p, dim in itertools.product(sizes, dtypes, p_values, dim_options):
    try:
        x = torch.randn(size, device=device, dtype=dtype)
    except RuntimeError:
        continue

    try:
        t = run_benchmark(x, p, dim)
    except RuntimeError:
        continue

    results.append((size, dtype, p, dim, t))

print(f"{'size'} {'dtype'} {'p'} {'dim'} {'median,(s)'}")
print()

for size, dtype, p, dim, t in results:
    print(f"{str(size).replace(' ','')} {str(dtype).split('.')[-1]} {str(p)} {str(dim)} {t:12.6e}")
```

</details>

Results:

<details><summary>Click to expand</summary>

size | dtype | p | dim | runtime before, s | runtime after, s | speedup
-- | -- | -- | -- | -- | -- | --
  |   |   |   |   |   |  
(10,) | float32 | 1.5 | None | 1.28E-05 | 9.26E-06 | 1.38
(10,) | float32 | 1.5 | 0 | 1.25E-05 | 8.80E-06 | 1.42
(10,) | float32 | 1.5 | -1 | 1.26E-05 | 9.00E-06 | 1.40
(10,) | float32 | 0 | None | 1.26E-05 | 7.57E-06 | 1.67
(10,) | float32 | 0 | 0 | 1.25E-05 | 7.30E-06 | 1.71
(10,) | float32 | 0 | -1 | 1.26E-05 | 7.60E-06 | 1.66
(10,) | float16 | 1.5 | None | 1.27E-05 | 7.74E-06 | 1.64
(10,) | float16 | 1.5 | 0 | 1.25E-05 | 7.59E-06 | 1.64
(10,) | float16 | 1.5 | -1 | 1.26E-05 | 8.04E-06 | 1.56
(10,) | float16 | 0 | None | 1.25E-05 | 7.45E-06 | 1.68
(10,) | float16 | 0 | 0 | 1.25E-05 | 6.65E-06 | 1.88
(10,) | float16 | 0 | -1 | 1.25E-05 | 6.64E-06 | 1.88
(10,) | bfloat16 | 1.5 | None | 1.27E-05 | 7.94E-06 | 1.60
(10,) | bfloat16 | 1.5 | 0 | 1.25E-05 | 6.15E-06 | 2.04
(10,) | bfloat16 | 1.5 | -1 | 1.25E-05 | 5.69E-06 | 2.20
(10,) | bfloat16 | 0 | None | 1.27E-05 | 4.94E-06 | 2.57
(10,) | bfloat16 | 0 | 0 | 1.27E-05 | 4.69E-06 | 2.71
(10,) | bfloat16 | 0 | -1 | 1.25E-05 | 4.61E-06 | 2.72
(10,10) | float32 | 1.5 | None | 1.34E-05 | 6.40E-06 | 2.09
(10,10) | float32 | 1.5 | 0 | 1.27E-05 | 5.73E-06 | 2.22
(10,10) | float32 | 1.5 | -1 | 1.28E-05 | 5.53E-06 | 2.32
(10,10) | float32 | 0 | None | 1.36E-05 | 5.48E-06 | 2.48
(10,10) | float32 | 0 | 0 | 1.24E-05 | 5.22E-06 | 2.37
(10,10) | float32 | 0 | -1 | 1.25E-05 | 4.98E-06 | 2.51
(10,10) | float16 | 1.5 | None | 1.32E-05 | 6.43E-06 | 2.06
(10,10) | float16 | 1.5 | 0 | 1.28E-05 | 5.71E-06 | 2.23
(10,10) | float16 | 1.5 | -1 | 1.27E-05 | 5.92E-06 | 2.15
(10,10) | float16 | 0 | None | 1.33E-05 | 5.22E-06 | 2.55
(10,10) | float16 | 0 | 0 | 1.26E-05 | 5.12E-06 | 2.46
(10,10) | float16 | 0 | -1 | 1.28E-05 | 4.69E-06 | 2.73
(10,10) | bfloat16 | 1.5 | None | 1.35E-05 | 6.46E-06 | 2.10
(10,10) | bfloat16 | 1.5 | 0 | 1.26E-05 | 5.79E-06 | 2.18
(10,10) | bfloat16 | 1.5 | -1 | 1.29E-05 | 5.80E-06 | 2.22
(10,10) | bfloat16 | 0 | None | 1.34E-05 | 5.29E-06 | 2.54
(10,10) | bfloat16 | 0 | 0 | 1.25E-05 | 4.93E-06 | 2.55
(10,10) | bfloat16 | 0 | -1 | 1.26E-05 | 5.06E-06 | 2.49
(1024,) | float32 | 1.5 | None | 1.30E-05 | 7.00E-06 | 1.86
(1024,) | float32 | 1.5 | 0 | 1.27E-05 | 6.85E-06 | 1.86
(1024,) | float32 | 1.5 | -1 | 1.29E-05 | 6.92E-06 | 1.87
(1024,) | float32 | 0 | None | 1.28E-05 | 5.77E-06 | 2.22
(1024,) | float32 | 0 | 0 | 1.27E-05 | 5.69E-06 | 2.23
(1024,) | float32 | 0 | -1 | 1.28E-05 | 5.72E-06 | 2.23
(1024,) | float16 | 1.5 | None | 1.33E-05 | 6.93E-06 | 1.92
(1024,) | float16 | 1.5 | 0 | 1.31E-05 | 6.95E-06 | 1.89
(1024,) | float16 | 1.5 | -1 | 1.30E-05 | 6.90E-06 | 1.89
(1024,) | float16 | 0 | None | 1.31E-05 | 5.71E-06 | 2.29
(1024,) | float16 | 0 | 0 | 1.28E-05 | 5.56E-06 | 2.29
(1024,) | float16 | 0 | -1 | 1.27E-05 | 5.53E-06 | 2.30
(1024,) | bfloat16 | 1.5 | None | 1.31E-05 | 6.98E-06 | 1.88
(1024,) | bfloat16 | 1.5 | 0 | 1.27E-05 | 6.91E-06 | 1.83
(1024,) | bfloat16 | 1.5 | -1 | 1.28E-05 | 6.74E-06 | 1.89
(1024,) | bfloat16 | 0 | None | 1.31E-05 | 5.79E-06 | 2.26
(1024,) | bfloat16 | 0 | 0 | 1.26E-05 | 5.62E-06 | 2.24
(1024,) | bfloat16 | 0 | -1 | 1.26E-05 | 5.73E-06 | 2.20
(1024,10) | float32 | 1.5 | None | 1.50E-05 | 2.70E-05 | 0.56
(1024,10) | float32 | 1.5 | 0 | 2.09E-05 | 7.47E-06 | 2.80
(1024,10) | float32 | 1.5 | -1 | 1.32E-05 | 1.45E-05 | 0.91
(1024,10) | float32 | 0 | None | 1.52E-05 | 2.08E-05 | 0.73
(1024,10) | float32 | 0 | 0 | 1.32E-05 | 6.60E-06 | 2.00
(1024,10) | float32 | 0 | -1 | 1.27E-05 | 1.02E-05 | 1.25
(1024,10) | float16 | 1.5 | None | 1.49E-05 | 2.71E-05 | 0.55
(1024,10) | float16 | 1.5 | 0 | 2.05E-05 | 7.51E-06 | 2.72
(1024,10) | float16 | 1.5 | -1 | 1.29E-05 | 1.45E-05 | 0.89
(1024,10) | float16 | 0 | None | 1.47E-05 | 2.08E-05 | 0.70
(1024,10) | float16 | 0 | 0 | 1.27E-05 | 6.22E-06 | 2.04
(1024,10) | float16 | 0 | -1 | 1.27E-05 | 1.00E-05 | 1.27
(1024,10) | bfloat16 | 1.5 | None | 1.48E-05 | 2.75E-05 | 0.54
(1024,10) | bfloat16 | 1.5 | 0 | 2.06E-05 | 7.55E-06 | 2.73
(1024,10) | bfloat16 | 1.5 | -1 | 1.29E-05 | 1.46E-05 | 0.88
(1024,10) | bfloat16 | 0 | None | 1.48E-05 | 2.09E-05 | 0.71
(1024,10) | bfloat16 | 0 | 0 | 1.28E-05 | 6.21E-06 | 2.06
(1024,10) | bfloat16 | 0 | -1 | 1.27E-05 | 1.01E-05 | 1.25
(10,10,10) | float32 | 1.5 | None | 1.41E-05 | 8.30E-06 | 1.69
(10,10,10) | float32 | 1.5 | 0 | 1.29E-05 | 6.57E-06 | 1.97
(10,10,10) | float32 | 1.5 | -1 | 1.29E-05 | 6.50E-06 | 1.99
(10,10,10) | float32 | 0 | None | 1.41E-05 | 7.39E-06 | 1.91
(10,10,10) | float32 | 0 | 0 | 1.29E-05 | 5.52E-06 | 2.34
(10,10,10) | float32 | 0 | -1 | 1.29E-05 | 5.42E-06 | 2.38
(10,10,10) | float16 | 1.5 | None | 1.41E-05 | 7.85E-06 | 1.79
(10,10,10) | float16 | 1.5 | 0 | 1.31E-05 | 6.49E-06 | 2.02
(10,10,10) | float16 | 1.5 | -1 | 1.32E-05 | 6.60E-06 | 2.00
(10,10,10) | float16 | 0 | None | 1.42E-05 | 7.34E-06 | 1.93
(10,10,10) | float16 | 0 | 0 | 1.30E-05 | 5.36E-06 | 2.42
(10,10,10) | float16 | 0 | -1 | 1.29E-05 | 5.49E-06 | 2.35
(10,10,10) | bfloat16 | 1.5 | None | 1.41E-05 | 8.40E-06 | 1.68
(10,10,10) | bfloat16 | 1.5 | 0 | 1.31E-05 | 6.57E-06 | 1.99
(10,10,10) | bfloat16 | 1.5 | -1 | 1.32E-05 | 6.51E-06 | 2.02
(10,10,10) | bfloat16 | 0 | None | 1.41E-05 | 6.96E-06 | 2.03
(10,10,10) | bfloat16 | 0 | 0 | 1.31E-05 | 5.45E-06 | 2.39
(10,10,10) | bfloat16 | 0 | -1 | 1.32E-05 | 5.54E-06 | 2.38

</details>

In most of the cases, performance improved by around 2x. In some cases, when the input is size (1024, 10) and the reduction dim is -1, the performance was worse, with a speedup as low as 54%.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: aten/src/ATen/native/mps/kernels/ReduceOps.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+#include <c10/metal/common.h>
+
+#define MAX_THREADGROUP_SIZE static_cast<uint32_t>(256)
```
Comment (malfet) [inline aten/src/ATen/native/mps/kernels/ReduceOps.h]: Err, where is this coming from? Usually MaxThreadgroup size is 1024

File: aten/src/ATen/native/mps/kernels/ReduceOps.metal [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    } else {
+      auto index_in_dim = reduction_element_idx % input_dim_size;
+      reduction_element_idx /= input_dim_size;
+      input_offset += index_in_dim * params.input_strides[dim];
+    }
+  }
+  return input_offset;
+}
+
+template <typename TI, typename TC, typename TO>
```
Comment (malfet) [inline aten/src/ATen/native/mps/kernels/ReduceOps.metal]: Mind adding a quick comment what `TC` is (and why it's not the same as TA)

**Review by malfet — APPROVED** [review]
LGTM, but please add a benchmark in PR description

**Review by malfet — COMMENTED** [review]
Also, why not replace `norm_out_mps` with `norm_out` and register a dispatch to share error checking code better? (And same for `linalg_vector_norm_out_mps`)

File: aten/src/ATen/native/mps/kernels/ReduceOps.h (hunk 1 above)
Comment (kurtamohler) [reply aten/src/ATen/native/mps/kernels/ReduceOps.h]: I'll change to 1024. I wasn't 100% sure if it's 1024 for all platforms so I was just being conservative until I knew better

**Comment by kurtamohler** [comment]
I've pushed a fairly significant change, so it may be worth another review.

I added a benchmark to the description. Let me know if you'd like me to add more cases. Most of the tested cases are improved, and some have a slowdown. I'm not totally sure how to improve those, but I can spend some more time looking into it if you'd like

File: aten/src/ATen/native/mps/kernels/ReduceOps.metal [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+struct norm_abs_functor {
+  template <typename T, enable_if_t<!is_complex_v<T>, bool> = true>
+  inline T operator()(const T x) {
+    return static_cast<T>(::precise::abs(x));
+  }
+
+  template <typename T, enable_if_t<is_complex_v<T>, bool> = true>
+  inline float operator()(const T x) {
+    const auto abs_2 = ::precise::abs(float2(x));
```
Comment (malfet) [inline aten/src/ATen/native/mps/kernels/ReduceOps.metal:18]: You don't need abs here, do you? as `hypot` does sqare anyway

**Review by malfet — APPROVED** [review]

File: aten/src/ATen/native/mps/kernels/ReduceOps.metal (hunk 3 above)
Comment (kurtamohler) [reply aten/src/ATen/native/mps/kernels/ReduceOps.metal:18]: The abs call is necessary. `hypot` requires that the inputs are both non-negative

Comment (kurtamohler) [reply aten/src/ATen/native/mps/kernels/ReduceOps.metal:18]: Perhaps I should remove that requirement from `hypot`. I can do that in a follow-up
